lib/charms/grafana_agent/v0/cos_agent.py

```diff
@@ -251,7 +251,8 @@ class _MetricsEndpointDict(TypedDict):
 
 class CosAgentProviderUnitData(pydantic.BaseModel):
     """Unit databag model for `cos-agent` relation."""
-
+    class Config:
+        arbitrary_types_allowed = True
```

Comment:
What error do we get without this config, but with pydantic>2?

Reply:
I get the below
```
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'cosl.grafana_dashboard.GrafanaDashboard'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

I think this is because of how the `GrafanaDashboard` is defined, possibly we can open a seperate issue to get this migrated to a pydantic custom type?

```
class GrafanaDashboard(str):
    """GrafanaDashboard represents an actual dashboard in Grafana.

    The class is used to compress and encode, or decompress and decode,
    Grafana Dashboards in JSON format using LZMA.
    """

    @staticmethod
    def _serialize(raw_json: Union[str, bytes]) -> "GrafanaDashboard":
        if not isinstance(raw_json, bytes):
            raw_json = raw_json.encode("utf-8")
        encoded = base64.b64encode(lzma.compress(raw_json)).decode("utf-8")
        return GrafanaDashboard(encoded)

    def _deserialize(self) -> Dict:
        try:
            raw = lzma.decompress(base64.b64decode(self.encode("utf-8"))).decode()
            return json.loads(raw)
        except json.decoder.JSONDecodeError as e:
            logger.error("Invalid Dashboard format: %s", e)
            return {}

    def __repr__(self):
        """Return string representation of self."""
        return "<GrafanaDashboard>"

```

Reply:
yeah in pydantic v2 you can't use arbitrary classes. Custom type is the way to go, or declare it as str and do serialization/deserialization outside of the model

Reply:
Captures in https://github.com/canonical/cos-lib/issues/23.